### PR TITLE
Save profile pic content to a cached file

### LINKED ISSUE #28: Save profile pic content to a cached file
- [x] create cache/images folders during app startup (exluded from git)
- [x] convert the base64 string loaded from Microsoft to bytes
- [x] save theses bytes into a jpeg file within the cache folder, name it like the id of the user.
- [x] create an addional route, where you can request the image from the cache folder
- [x] readup about http cache headers. Set a reasonable time for the image to be cached in the browser (e.g. 30 mins)
- [x] Bonus: Learn about "cache busting". Add a way to force the browser to load a new image and ignore the cached one.
- [x] Bonus: resize the image into a smaller variant. Save it e.g. under cache/thumbnails/…  that can be used, if only smaller pictures need to be shown.

### PR DESCRIPTION
**Pull request by curiosdevcookie**
- create cache/images folders during app startup.
- jpeg file with user_id as name.
- Add image request route.
- Set cache header to 1800 max-age.

Closes #28.

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    imageType === "images" ? cacheImagesFolderPath : cacheThumbnailsFolderPath;
+
+  const fileName =
+    imageType === "images"
+      ? `${userId}.jpeg`
+      : imageType === "thumbnails"
+      ? `${userId}.jpeg`
+      : null;
+
+  if (!fileName) {
```
Comment (phaus) [inline src/utils.ts:56]: I think if the fileName is null, in line 63, no file will be found?

File: src/app.ts [hunk 2]
```diff
@@ -66,6 +69,17 @@ app.use(logoutRoute);
...
+app.use(
+  "/cache/images",
+  express.static(path.join(__dirname, "..", "cache", "images"))
+);
+
+app.use(
+  "/cache/thumbnails",
+  express.static(path.join(__dirname, "..", "cache", "thumbnails"))
+);
+
```
[the remark this answers is not in the document]
Comment (curiosdevcookie) [reply src/app.ts:82]: True!

File: src/routes/mainRoutes.ts [hunk 3]
```diff
@@ -11,24 +11,48 @@ router.get("/", (req, res) => {
...
+  );
 });
 
 router.get("/profile", (req, res) => {
   if (!req.isAuthenticated()) {
     res.redirect("/");
   } else {
     const authenticatedUser = req.user;
-    res.render("profile", { authenticatedUser });
+    const cacheBuster = Date.now();
```
[the remark this answers is not in the document]
Comment (curiosdevcookie) [reply src/routes/mainRoutes.ts:36]: Check!

File: src/utils.ts (hunk 1 above)
Comment (curiosdevcookie) [reply src/utils.ts:56]: Yes?

Comment (phaus) [reply src/utils.ts:56]: Then you maybe don't need the null check before? ;-)

File: package.json [hunk 4]
```diff
@@ -6,7 +6,8 @@
   "main": "dist/app.js",
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "build": "tsc",
+    "build": "tsc && npm run copy",
+    "copy": "cp config.json dist/",
```
Comment (phaus) [inline package.json:10]: This looks strange. 

COPY ../config.json . 

should work as well :-)

Comment (curiosdevcookie) [reply package.json:10]: Not sure what you mean

**Review by phaus — APPROVED** [review]

**Pull request merged**